go/tools/gazelle/rules/generator.go

```diff
@@ -227,10 +237,14 @@ func (g *generator) generateXTest(pkg *packages.Package, library string) bf.Expr
 	}
 
 	name := g.l.TestLabel(pkg.Rel, true).Name
-	return g.generateRule(pkg.Rel, "go_test", name, "", "", pkg.HasTestdata, pkg.XTest)
+	var rundir string
+	if g.c.StructureMode == config.FlatMode {
+		rundir = pkg.Rel
+	}
+	return g.generateRule(pkg.Rel, "go_test", name, "", "", "", rundir, pkg.HasTestdata, pkg.XTest)
 }
 
-func (g *generator) generateRule(pkgRel, kind, name, visibility, library string, hasTestdata bool, target packages.Target) bf.Expr {
+func (g *generator) generateRule(pkgRel, kind, name, visibility, library, importpath, rundir string, hasTestdata bool, target packages.Target) bf.Expr {
```

Comment:
Completely agree. I drafted another commit this morning to fix this, but I'll need to merge it with this before sending it.

The plan is replace `generateRule` with a `commonAttrs` methods that returns `[]keyvalue` and only deals with attributes used by all rules. Other methods will add their own attributes, then call `newRule` directly.